invenio_drafts_resources/records/systemfields/versions.py

```diff
@@ -0,0 +1,259 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
+#
+# Invenio-Drafts-Resources is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Versions field."""
+
+import uuid
+from copy import copy
+
+from invenio_db import db
+from invenio_records.systemfields import RelatedModelField
+from invenio_records.systemfields.base import SystemField
+from sqlalchemy import inspect
+from sqlalchemy.exc import IntegrityError
+
+
+def uuid_or_none(val):
+    """Convert string to UUID object."""
+    if val is not None and not isinstance(val, uuid.UUID):
+        return uuid.UUID(val)
+    return val
+
+
+class VersionsManager:
+    """Versions state manager."""
+
+    def __init__(self, record, dump=None):
+        """Initialize the versions manager."""
+        self._record = record
+        self._state = None
+        if dump is not None:
+            self.load(dump)
+
+    def copy_to(self, record):
+        """Create a copy of the version manager and set on another record."""
+        record.model.parent_index = self.parent_index
+        version_manager = self.__class__(record)
+        version_manager._state = self._state
+        return version_manager
+
+    #
+    # Record managed attributes
+    #
+    @property
+    def model_cls(self):
+        """Get versions state management model."""
+        return self._record.versions_model_cls
+
+    @property
+    def parent_id(self):
+        """Get versions state management model."""
+        return self._record.model.parent_id
+
+    @property
+    def parent_index(self):
+        """Get the parent index."""
+        return self._record.model.parent_index
+
+    #
+    # Parent managed attributes
+    #
+    @property
+    def latest_id(self):
+        """The id of the latest published record/draft."""
+        return self.state().latest_id
+
+    @property
+    def latest_index(self):
+        """The index of the latest published record/draft."""
+        return self.state().latest_index
+
+    @property
+    def next_draft_id(self):
+        """The id of the next draft (and record)."""
+        return self.state().next_draft_id
+
+    #
+    # Computed attributes
+    #
+    @property
+    def is_latest(self):
+        """Check if the record/draft id is the latest published record id."""
+        return self.latest_id == self._record.id
+
+    @property
+    def is_latest_draft(self):
+        """Check if the record/draft id is the latest draft id."""
+        if self.next_draft_id:
+            return self.next_draft_id == self._record.id
+        else:
+            return self.latest_id == self._record.id
```

Comment:
It's related to the list uploads - i.e. when we search drafts we search over both records and drafts. 

The list uploads page can show only the latest version. This means that we apply a filter that includes only records/drafts that have ``record.is_latest_draft == True``. 

There's 3 cases where above condition has to return True:

1) The latest published record has only a record (i.e. it's not being edited) - we display the record
2) The latest published record has a record and draft (i.e. it's being edited) - we display the draft
3) A next draft exists (i.e. the we created a new version, but didn't publish yet) - we display the next draft.

The if-part is related to (3), then else-part is related to (1) and (2).

Reply:
So list uploads when "all versions" toggle switch is off shows:

```
Draft A (draft of an unpublished yet record)
Record B (published record with no drafts)
Draft C (draft of a published record)
Draft D2 (draft of the new -2nd- version of record D)
```

When toggle switch is on:

```
Draft A (draft of an unpublished yet record)
Record B (published record with no drafts)
Draft C (draft of a published record)
Draft D2 (draft of the new -2nd- version of record D)
Record D (published record for first version of record D) 
```
?


Like in overall review message, what is not clear to me is whether `record.versions.is_latest_draft` means:

a) the latest version of this record is a draft 
b) the current record is the latest version and there is a draft of it

I think it's a) (really I think it's a c) I don't grasp yet :smile: ). Naming could help:

a) -> `record.versions.latest_is_draft` (read as: the record's latest version is draft)
b) -> `record.versions.has_latest_draft` (read as: the record has the latest draft)

(same naming difficulty because is `versions` the subject of these methods or is it just a namespace? Plus record is overloaded, but that's another can of worms :sweat_smile: )